flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java

```diff
@@ -0,0 +1,672 @@
+import org.apache.flink.streaming.runtime.partitioner.ForwardPartitioner;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nullable;
+import java.util.*;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.streaming.api.graph.StreamingJobGraphGenerator.*;
+
+/** Default implementation for {@link AdaptiveGraphGenerator}. */
+@Internal
+public class AdaptiveGraphManager implements AdaptiveGraphGenerator {
+
+    private final StreamGraph streamGraph;
+
+    private final JobGraph jobGraph;
+
+    private final StreamGraphHasher defaultStreamGraphHasher;
+
+    private final List<StreamGraphHasher> legacyStreamGraphHasher;
+
+    private final Executor serializationExecutor;
+
+    private final AtomicInteger vertexIndexId;
+
+    private final StreamGraphContext streamGraphContext;
+
+    private final Map<Integer, byte[]> hashes;
+
+    private final List<Map<Integer, byte[]>> legacyHashes;
+
+    // Records the id of stream node which job vertex is created.
+    private final Map<Integer, Integer> frozenNodeToStartNodeMap;
+
+    // When the downstream vertex is not created, we need to cache the output.
+    private final Map<Integer, Map<StreamEdge, NonChainedOutput>> intermediateOutputsCaches;
+
+    // Records the id of the stream node that produces the IntermediateDataSet.
+    private final Map<IntermediateDataSetID, Integer> intermediateDataSetIdToProducerMap;
+
+    // Records the ids of the start and end nodes for chained groups in the StreamNodeForwardGroup.
+    // When stream edge's partitioner is modified to forward, we need get forward groups by source
+    // and target node id.
+    private final Map<Integer, StreamNodeForwardGroup> startAndEndNodeIdToForwardGroupMap;
+
+    // Records the chain info that is not ready to create the job vertex, the key is the start node
+    // in this chain.
+    private final Map<Integer, OperatorChainInfo> pendingChainEntryPoints;
+
+    // The value is the stream node ids belonging to that job vertex.
+    private final Map<JobVertexID, Integer> jobVertexToStartNodeMap;
+
+    // The value is the stream node ids belonging to that job vertex.
+    private final Map<JobVertexID, List<Integer>> jobVertexToChainedStreamNodeIdsMap;
+
+    // We need cache all job vertices to create JobEdge for downstream vertex.
+    private final Map<Integer, JobVertex> jobVerticesCache;
+
+    // Records the ID of the job vertex that has completed execution.
+    private final Set<JobVertexID> finishedJobVertices;
+
+    private final AtomicBoolean hasHybridResultPartition;
+
+    public AdaptiveGraphManager(
+            ClassLoader userClassloader,
+            StreamGraph streamGraph,
+            Executor serializationExecutor,
+            @Nullable JobID jobId) {
```

Comment:
I think the job id should be obtained from the StreamGraph instead of being passed as a parameter.